node_package/src/reactHydrateOrRender.ts

```diff
@@ -0,0 +1,44 @@
+import type { ReactElement } from 'react';
+import ReactDOM from 'react-dom';
+import type { RenderReturnType } from './types';
+
+type HydrateOrRenderType = (domNode: Element, reactElement: ReactElement) => RenderReturnType;
+const supportsReactCreateRoot = parseInt(ReactDOM.version.split('.')[0], 10) >= 18;
+
+// TODO: once React dependency is updated to >= 18, we can remove this and just
+// import ReactDOM from 'react-dom/client';
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+let reactDomClient: any;
+if (supportsReactCreateRoot) {
+  // eslint-disable-next-line camelcase
+  if (__webpack_require__) {
```

Comment:
Wonder if this is the magic thing that was missing in the other attempts!

Reply:
No, it isn't :) If I understand correctly (Webpack documentation is bad), https://github.com/shakacode/react_on_rails/pull/1448 was actually close to working, but Webpack needs a base directory for dynamic imports/requires. If it had `require(`react-dom/${supportsReactCreateRoot ? 'client' : 'index'}`);`, it would recognize `react_dom` as the base directory and work.

Reply:
> Don't have React 18 app on hand to test though, if we have test cases that use it, it's good enough

It doesn't work after all. Maybe the generated examples don't use Webpack?

Reply:
How about if we just use a different import for react 18 with react on rails?

Check source for react-rails?

Google to see what other libraries do which have react as a dependency?

Reply:
Some of the libraries I've seen that got direct dependency react-dom seem to have just released major versions that only support React 18 and left it at that. For all intents and purposes, React 18 components not initialized with the new root API should function the same as they do in React 17.

I do feel there's bit more at play here, think I've noted some other changes that might be required at one of the other PRs trying to tackle this.